File: lang/javascript/sdk/src/discovery/hcl.js
"use strict";

/**
 * Minimal Terraform HCL parser for discovering AWS resources in .tf files.
 *
 * Supports the subset of HCL needed for common AWS resource types, including
 * heredoc strings (<<MARKER ... MARKER) used for multi-line JSON definitions.
 */

const fs = require("fs");
const path = require("path");

const RESOURCE_HEADER = /resource\s+"(?<type>[^"]+)"\s+"(?<logical>[^"]+)"\s*\{/;
const ATTR_STR = /^\s*(?<key>\w+)\s*=\s*"(?<value>[^"]*)"/;
const ATTR_BARE = /^\s*(?<key>\w+)\s*=\s*(?<value>\S+)/;
const ATTR_HEREDOC = /^\s*(?<key>\w+)\s*=\s*<<(?<marker>\w+)\s*$/;

/**
 * Parse all .tf files in `projectDir` and return a resource spec object with
 * discovered AWS resources ready for use with LwsSession.
 *
 * @param {string} projectDir
 * @returns {{ tables?: Array, queues?: Array, buckets?: Array, topics?: Array, stateMachines?: Array, parameters?: Array, secrets?: Array }}
 */
function discoverHcl(projectDir) {
  const tfFiles = [];

  function findTfFiles(dir) {
    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
      if (entry.isDirectory() && !entry.name.startsWith(".")) {
        findTfFiles(path.join(dir, entry.name));
      } else if (entry.name.endsWith(".tf")) {
        tfFiles.push(path.join(dir, entry.name));
      }
    }
  }

  findTfFiles(projectDir);

  if (tfFiles.length === 0) {
    throw new Error(
      `No .tf files found in ${projectDir}. ` +
        "Make sure you point to the directory containing your Terraform files."
    );
  }

  const resources = [];
  for (const tfFile of tfFiles) {
    resources.push(...parseTfFile(tfFile));
  }

  return classifyResources(resources);
}

function parseTfFile(filePath) {
  const text = fs.readFileSync(filePath, "utf-8");
  const lines = text.split("\n");
  const results = [];
  let i = 0;

  while (i < lines.length) {
    const m = RESOURCE_HEADER.exec(lines[i]);
    if (m?.groups) {
      const { type, logical } = m.groups;
      const [attrs, nextI] = collectBlock(lines, i + 1);
      results.push({ type, logical, attrs });
      i = nextI;
    } else {
      i++;
    }
  }

  return results;
}

function collectBlock(lines, start) {
  const attrs = {};
  let depth = 1;
  let i = start;

  while (i < lines.length && depth > 0) {
    const line = lines[i];

    // Handle heredoc (key = <<MARKER ... MARKER) before adjusting depth so
    // JSON content inside the heredoc does not affect brace counting.
    if (depth === 1) {
      const mHeredoc = ATTR_HEREDOC.exec(line);
      if (mHeredoc?.groups) {
        const { key, marker } = mHeredoc.groups;
        i++;
        const heredocLines = [];
        while (i < lines.length && lines[i].trimEnd() !== marker) {
          heredocLines.push(lines[i]);
          i++;
        }
        attrs[key] = heredocLines.join("\n");
        i++; // skip the closing marker line
        continue;
      }
    }

    const opens = (line.match(/\{/g) ?? []).length;
    const closes = (line.match(/\}/g) ?? []).length;
    depth += opens - closes;

    if (depth > 1) {
      i++;
      continue;
    }
    if (depth === 0) break;

    const mStr = ATTR_STR.exec(line);
    if (mStr?.groups) {
      attrs[mStr.groups.key] = mStr.groups.value;
    } else {
      const mBare = ATTR_BARE.exec(line);
      if (mBare?.groups) {
        attrs[mBare.groups.key] = mBare.groups.value;
      }
    }
    i++;
  }

  return [attrs, i + 1];
}

function classifyResources(resources) {
  const spec = {
    tables: [],
    queues: [],
    buckets: [],
    topics: [],
    stateMachines: [],
    parameters: [],
    secrets: [],
  };

  for (const { type, attrs } of resources) {
    switch (type) {
      case "aws_dynamodb_table": {
        const table = buildDynamoDBTable(attrs);
        if (table) spec.tables.push(table);
        break;
      }
      case "aws_sqs_queue": {
        const queue = buildSqsQueue(attrs);
        if (queue) spec.queues.push(queue);
        break;
      }
      case "aws_s3_bucket": {
        const bucket = buildS3Bucket(attrs);
        if (bucket) spec.buckets.push(bucket);
        break;
      }
      case "aws_sns_topic": {
        const topic = buildSnsTopic(attrs);
        if (topic) spec.topics.push(topic);
        break;
      }
      case "aws_sfn_state_machine": {
        const sm = buildStateMachine(attrs);
        if (sm) spec.stateMachines.push(sm);
        break;
      }
      case "aws_ssm_parameter": {
        const param = buildSsmParameter(attrs);
        if (param) spec.parameters.push(param);
        break;
      }
      case "aws_secretsmanager_secret": {
        const secret = buildSecretsManagerSecret(attrs);
        if (secret) spec.secrets.push(secret);
        break;
      }
    }
  }

  return spec;
}

function buildDynamoDBTable(attrs) {
  if (!attrs["name"]) return null;
  const table = {
    name: attrs["name"],
    partitionKey: attrs["hash_key"] ?? "id",
  };
  if (attrs["range_key"]) {
    table.sortKey = attrs["range_key"];
  }
  return table;
}

function buildSqsQueue(attrs) {
  if (!attrs["name"]) return null;
  const queue = { name: attrs["name"] };
  if (attrs["fifo_queue"] === "true") {
    queue.isFifo = true;
  }
  return queue;
}

function buildS3Bucket(attrs) {
  const name = attrs["bucket"] ?? attrs["name"];
  if (!name) return null;
  return { name };
}

function buildSnsTopic(attrs) {
  if (!attrs["name"]) return null;
  return { name: attrs["name"] };
}

function buildStateMachine(attrs) {
  if (!attrs["name"]) return null;
  return {
    name: attrs["name"],
    definition: attrs["definition"] ?? "{}",
    roleArn: attrs["role_arn"],
  };
}

function buildSsmParameter(attrs) {
  if (!attrs["name"]) return null;
  const param = { name: attrs["name"] };
  if (attrs["value"]) param.value = attrs["value"];
  if (attrs["type"]) param.type = attrs["type"];
  return param;
}

function buildSecretsManagerSecret(attrs) {
  if (!attrs["name"]) return null;
  const secret = { name: attrs["name"] };
  if (attrs["description"]) secret.description = attrs["description"];
  return secret;
}

module.exports = { discoverHcl };

File: lang/python/sdk/src/lws_testing/_discovery/hcl.py
"""Discover AWS resources from Terraform HCL (.tf) files."""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any

# Matches:  resource "aws_dynamodb_table" "logical_name" {
_RESOURCE_HEADER = re.compile(
    r'resource\s+"(?P<type>[^"]+)"\s+"(?P<logical>[^"]+)"\s*\{',
)
# Matches a simple  key = "value"  assignment
_ATTR_STR = re.compile(r'^\s*(?P<key>\w+)\s*=\s*"(?P<value>[^"]*)"')
# Matches a simple  key = value  (unquoted booleans / numbers)
_ATTR_BARE = re.compile(r"^\s*(?P<key>\w+)\s*=\s*(?P<value>\S+)")
# Matches a heredoc opener:  key = <<MARKER
_ATTR_HEREDOC = re.compile(r"^\s*(?P<key>\w+)\s*=\s*<<(?P<marker>\w+)\s*$")

# Maps Terraform resource types to (collection_key, builder_function)
_DISPATCH: dict[str, tuple[str, Any]] = {
    "aws_dynamodb_table": ("tables", "_build_table"),
    "aws_sqs_queue": ("queues", "_build_queue"),
    "aws_s3_bucket": ("buckets", "_build_bucket"),
    "aws_sns_topic": ("topics", "_build_topic"),
    "aws_sfn_state_machine": ("state_machines", "_build_state_machine"),
    "aws_ssm_parameter": ("parameters", "_build_parameter"),
    "aws_secretsmanager_secret": ("secrets", "_build_secret"),
}


def discover(project_dir: Path) -> dict[str, Any]:
    """Parse ``.tf`` files in *project_dir* and return a resource spec dict.

    Returns a dict compatible with :class:`~lws_testing.LwsSession` kwargs:
    ``tables``, ``queues``, ``buckets``, ``topics``, ``state_machines``,
    ``secrets``, ``parameters``.
    """
    tf_files = list(project_dir.rglob("*.tf"))
    if not tf_files:
        raise FileNotFoundError(
            f"No .tf files found in {project_dir}. "
            "Make sure you point to the directory containing your Terraform files."
        )

    resources: list[tuple[str, str, dict[str, str]]] = []
    for tf_file in tf_files:
        resources.extend(_parse_tf_file(tf_file))

    return _classify_resources(resources)


def _classify_resources(
    resources: list[tuple[str, str, dict[str, str]]],
) -> dict[str, list[Any]]:
    """Route parsed resources into typed collections."""
    result: dict[str, list[Any]] = {
        "tables": [],
        "queues": [],
        "buckets": [],
        "topics": [],
        "state_machines": [],
        "parameters": [],
        "secrets": [],
    }
    handlers = _get_handlers()
    for resource_type, _logical, attrs in resources:
        if resource_type not in _DISPATCH:
            continue
        key, handler_name = _DISPATCH[resource_type]
        item = handlers[handler_name](attrs)
        if item is not None:
            result[key].append(item)
    return result


def _get_handlers() -> dict[str, Any]:
    """Return a map of handler name to callable."""
    return {
        "_build_table": _build_table,
        "_build_queue": _build_queue,
        "_build_bucket": _build_bucket,
        "_build_topic": _build_topic,
        "_build_state_machine": _build_state_machine,
        "_build_parameter": _build_parameter,
        "_build_secret": _build_secret,
    }


def _parse_tf_file(path: Path) -> list[tuple[str, str, dict[str, str]]]:
    """Parse a single .tf file and return list of (resource_type, logical_name, attrs)."""
    text = path.read_text(encoding="utf-8")
    results: list[tuple[str, str, dict[str, str]]] = []
    lines = text.splitlines()
    i = 0
    while i < len(lines):
        m = _RESOURCE_HEADER.search(lines[i])
        if m:
            resource_type = m.group("type")
            logical_name = m.group("logical")
            attrs, i = _collect_block(lines, i + 1)
            results.append((resource_type, logical_name, attrs))
        else:
            i += 1
    return results


def _try_heredoc(lines: list[str], i: int) -> tuple[str, str, int] | None:
    """If line *i* is a heredoc opener, return *(key, value, next_i)*, else *None*."""
    m = _ATTR_HEREDOC.match(lines[i])
    if not m:
        return None
    key = m.group("key")
    marker = m.group("marker")
    i += 1
    body: list[str] = []
    while i < len(lines) and lines[i].rstrip() != marker:
        body.append(lines[i])
        i += 1
    return key, "\n".join(body), i + 1  # i+1 advances past the closing marker


def _extract_attr(line: str) -> tuple[str, str] | None:
    """Return *(key, value)* if *line* is a simple attribute assignment, else *None*."""
    m = _ATTR_STR.match(line)
    if m:
        return m.group("key"), m.group("value")
    m2 = _ATTR_BARE.match(line)
    if m2:
        return m2.group("key"), m2.group("value")
    return None


def _collect_block(lines: list[str], start: int) -> tuple[dict[str, str], int]:
    """Collect top-level key=value pairs until the matching closing brace."""
    attrs: dict[str, str] = {}
    depth = 1
    i = start
    while i < len(lines) and depth > 0:
        line = lines[i]

        # Check for heredoc before adjusting depth so JSON braces inside
        # the heredoc body do not corrupt the brace counter.
        if depth == 1:
            result = _try_heredoc(lines, i)
            if result is not None:
                key, value, i = result
                attrs[key] = value
                continue

        depth += line.count("{") - line.count("}")
        if depth > 1:
            i += 1
            continue
        if depth == 0:
            break
        pair = _extract_attr(line)
        if pair is not None:
            attrs[pair[0]] = pair[1]
        i += 1
    return attrs, i + 1


def _build_table(attrs: dict[str, str]) -> dict[str, Any] | None:
    """Build a table spec dict from DynamoDB resource attributes."""
    name = attrs.get("name")
    hash_key = attrs.get("hash_key")
    if not name or not hash_key:
        return None
    spec: dict[str, Any] = {"name": name, "partition_key": hash_key}
    if "range_key" in attrs:
        spec["sort_key"] = attrs["range_key"]
    return spec


def _build_queue(attrs: dict[str, str]) -> dict[str, Any] | None:
    """Build a queue spec dict from SQS resource attributes."""
    name = attrs.get("name")
    if not name:
        return None
    is_fifo = attrs.get("fifo_queue", "false").lower() in ("true", "1")
    spec: dict[str, Any] = {"name": name, "is_fifo": is_fifo}
    if "visibility_timeout_seconds" in attrs:
        try:
            spec["visibility_timeout"] = int(attrs["visibility_timeout_seconds"])
        except ValueError:
            pass
    return spec


def _build_bucket(attrs: dict[str, str]) -> str | None:
    """Build a bucket name from S3 resource attributes."""
    return attrs.get("bucket") or attrs.get("name")


def _build_topic(attrs: dict[str, str]) -> dict[str, str] | None:
    """Build a topic spec dict from SNS resource attributes."""
    name = attrs.get("name")
    return {"name": name} if name else None


def _build_state_machine(attrs: dict[str, str]) -> dict[str, Any] | None:
    """Build a state machine spec dict from Step Functions resource attributes."""
    name = attrs.get("name")
    if not name:
        return None
    return {
        "name": name,
        "definition": attrs.get("definition", "{}"),
        "role_arn": attrs.get("role_arn", ""),
    }


def _build_parameter(attrs: dict[str, str]) -> dict[str, Any] | None:
    """Build a parameter spec dict from SSM resource attributes."""
    name = attrs.get("name")
    if not name:
        return None
    return {
        "name": name,
        "value": attrs.get("value", ""),
        "type": attrs.get("type", "String"),
        "description": attrs.get("description", ""),
    }


def _build_secret(attrs: dict[str, str]) -> dict[str, Any] | None:
    """Build a secret spec dict from Secrets Manager resource attributes."""
    name = attrs.get("name")
    if not name:
        return None
    return {
        "name": name,
        "description": attrs.get("description", ""),
        "secret_string": "",
    }
